### ISSUE TITLE
South migrations models are cached by cacheops

### ISSUE THREAD

**Issue by pySilver** [Suor/django-cacheops#196]
Hi, 

I've just noticed something weird with my datamigrations while cacheops is enabled. Consider the example:

```
# Correct
In [25]: orm.User.mro()
Out[25]: [accounts.models.User, django.db.models.base.Model, object]

# Incorrect!
In [26]: orm.User.objects.first().__class__.mro()
Out[26]:
[accounts.models.User,
 django.contrib.auth.models.AbstractBaseUser,
 django.contrib.auth.models.PermissionsMixin,
 core.models.mixins.VersionMixin,
 core.models.mixins.ModelDiffMixin,
 django.db.models.base.Model,
 object]

# Correct
In [29]: orm.User.objects.nocache().first().__class__.mro()
Out[29]: [accounts.models.User, django.db.models.base.Model, object]
```

I do understand that south in not the current way of handling migrations, but still there are plenty of a projects running pre 1.7 django version. 

Probably this location is responsible for the problem:
https://github.com/Suor/django-cacheops/blob/2aa499740ae814e18905143aebe9d62ae552e2d9/cacheops/query.py#L376

**Comment by Suor** [comment]
First, I can't see if this is broken. Does migrations not run for you with cacheops and run without it?

Second, `.contribute_to_class()`, which you are refering to, is executed at the start of the program and hence, can't be affected by using `.nocache()`. 

Third, cacheops have been in use with south for a while (including myself), so there should be some component beyond these two involved.

**Comment by Suor** [comment]
My guess is that in first case you get a real User instance from cache and in second one you get an instance of South fake model. You can test this by looking at the class of an object returned.

**Comment by pySilver** [comment]
Nope, Im 100% confident its a cacheops that causes a problem. I've spend 5h+ investigating. 
You are right, that not really clear whats wrong here :) The problem is that cacheops causes class to be an original one and not the fake one that south creates. There will be no problem until you override a .save (you simply will not notice that). Moreover you wont notice that methods that are not supposed to be in a fake model are here (due to mro):

```
In [5]: u = orm.User.objects.first()

In [6]: u.__class__.mro()
Out[6]:
[accounts.models.User,
 django.contrib.auth.models.AbstractBaseUser,
 django.contrib.auth.models.PermissionsMixin,
 core.models.mixins.VersionMixin,
 core.models.mixins.ModelDiffMixin,
 django.db.models.base.Model,
 object]

In [9]: orm.User.objects.first()._format_phone_number
Out[9]: <bound method User._format_phone_number of <User: Foo Administrator>>
```

Compare with disabled cacheops:
```
In [8]: orm.User.objects.nocache().first()._format_phone_number
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/Users/Silver/Projects/****/migrations/0019_billing_field_renamed.pyc in <module>()
----> 1 orm.User.objects.nocache().first()._format_phone_number

AttributeError: 'User' object has no attribute '_format_phone_number'
```

**Comment by pySilver** [comment]
ofc correct behaviour is latter. south datamigrations are supposed to have *no access* to real model methods, so it guarantee migration would work in the future.

**Comment by Suor** [comment]
I agree that cacheops should not work on migration models and the way it works now that fake models are intermixed with normal ones is even worse. The question is how should we distinguish south fake models from normal ones.

Please try this in your migrations and does it return `True` or `False`:

```python
from south.orm import NoDryRunManager
isinstance(orm.User.objects, NoDryRunManager)
```

**Comment by pySilver** [comment]
@Suor yes, it is:

```python
In [1]: isinstance(orm.User.objects, NoDryRunManager)
Out[1]: True

In [2]: isinstance(orm.User.objects.nocache(), NoDryRunManager)
Out[2]: False

In [3]: isinstance(orm.User.objects.nocache, NoDryRunManager)
Out[3]: False
```

**Comment by Suor** [comment]
Should be fixed in 2.x branch. Can you confirm?

**Comment by pySilver** [comment]
I've just tested with 2.4.4 and it got even worse :) 

```
$ ./manage.py datamigration accounts test_migration
Traceback (most recent call last):
  File "./manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/core/management/__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/core/management/__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/core/management/__init__.py", line 272, in fetch_command
    klass = load_command_class(app_name, subcommand)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/core/management/__init__.py", line 75, in load_command_class
    module = import_module('%s.management.commands.%s' % (app_name, name))
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/utils/importlib.py", line 40, in import_module
    __import__(name)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/south/management/commands/__init__.py", line 10, in <module>
    import django.template.loaders.app_directories
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/template/loaders/app_directories.py", line 23, in <module>
    mod = import_module(app)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/utils/importlib.py", line 40, in import_module
    __import__(name)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/cacheops/__init__.py", line 32, in <module>
    install_cacheops()
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/funcy/flow.py", line 172, in wrapper
    return func(*args, **kwargs)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/cacheops/query.py", line 513, in install_cacheops
    for model in get_models(include_auto_created=True):
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/db/models/loading.py", line 232, in get_models
    self._populate()
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/db/models/loading.py", line 75, in _populate
    self.load_app(app_name, True)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/db/models/loading.py", line 99, in load_app
    models = import_module('%s.models' % app_name)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/utils/importlib.py", line 40, in import_module
    __import__(name)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/contrib/auth/models.py", line 361, in <module>
    class AbstractUser(AbstractBaseUser, PermissionsMixin):
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/db/models/base.py", line 144, in __new__
    new_class.add_to_class(obj_name, obj)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/db/models/base.py", line 264, in add_to_class
    value.contribute_to_class(cls, name)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/cacheops/query.py", line 375, in contribute_to_class
    if not model_is_fake(cls):
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/cacheops/conf.py", line 155, in model_is_fake
    or NoDryRunManager and isinstance(model.objects, NoDryRunManager)
  File "/Users/Silver/.virtualenvs/convertiser-server/lib/python2.7/site-packages/django/db/models/manager.py", line 261, in __get__
    self.model._meta.object_name,
AttributeError: Manager isn't available; AbstractUser is abstract
```

**Comment by pySilver** [comment]
missing check for abstract model, I suppose

**Comment by Suor** [comment]
Should be fixed in https://github.com/Suor/django-cacheops/commit/4f907bd4b755f16fb164f0fc0ea0126ed76a81ad.

**Comment by Suor** [comment]
Fix released in 2.4.5.
